client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelInitializer.java

```diff
@@ -86,8 +79,6 @@ protected void initChannel(Channel channel) throws Exception {
          channel.pipeline().addLast("idle-state-handler",
                new IdleStateHandler(0, 0, configuration.connectionPool().minEvictableIdleTime(), TimeUnit.MILLISECONDS));
       }
-      ChannelRecord channelRecord = new ChannelRecord(unresolvedAddress, channelPool);
```

Comment:
@rvansa Could you please review this and `ActivationFuture` changes a couple of lines below?

I believe we had a race condition here. The `ChannelRecord` entry was set in the `ChannelInitializer#initChannel` method but was used in `ActivationFuture#operationComplete`. I couldn't find any documentation that says that `initChannel` method is guaranteed to finish before a listener attached to channel activation is fired. As the matter of fact, I've been observing this failing pretty often when I added TLS/ALPN support which waits until the protocol is negotiated and *then* initializes the pipeline. This resulted in a `NullPointerException` thrown from `ChannelRecord.of(channel)`.

So I decided to move the `ChannelRecord` creation to the `ActivationFuture`. I ran the client testsuite a couple of times but it's so flaky that it's hard to detect if I broke anything or not. Please tell me what do you think about this change.

Reply:
> it's so flaky

What's flaky about it? We're not seeing such flakyness on CI.

Reply:
This part seems to rely on Netty internal ordering of events. When developing Single Port Client I noticed that it is possible that a connection upgrade finishes and the ActivationFuture is still "in flight". So in my opinion we are using a bit risky construct there which may change even in minor Netty upgrades.

Reply:
@slaskawi `initChannel` is called from `io.netty.channel.ChannelInitializer` when the channel is *registered*, while `ActivationFuture.operationCompleted` is called when the promise returned by `Bootstrap.connect()` is completed. This should be done only after the registration completes (see `Bootstrap.doResolveAndConnect()`)

Reply:
Not that I would have any issues with the way you did it instead, as long as it's non-blocking and testsuite passes.

Reply:
Ok, agreed. Thanks @rvansa for the input!